### PR TITLE
feat(migration): Add migration to add explicit Route/Routes type

### LINKED ISSUE #37469: router: stricter typing for Route.pathMatch
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/router

### Description

`Route.pathMatch` field is now typed as `string | undefined` but actually it can accept `prefix` or `full` as a valid value.
I want to restrict invalid values statically at compilation step with a string literal types.

### Describe the solution you'd like

Typing `pathMatch` as `pathMatch?: 'prefix' | 'full'`

### Describe alternatives you've considered

This is a breaking change because it may cause build error by narrowing the acceptable value (even if those already have caused runtime error).

### PR DESCRIPTION
**Pull request by atscott**
reviewer notes (specifically for the migration):
To make reviewing easier, the changes to the type are in a separate PR https://github.com/angular/angular/pull/45176
The approach used by the migration is not perfect but aims to migrate the large majority of use-cases. Shortcomings of this migration are:

* `Route`/`Routes` that need to be migrated are identified by having both "pathMatch" and "path" properties. There is a small risk of this being misidentified. Important to note, in all of g3, there were *0* misidentified objects (objects that were not truly used as an Angular `Route`)
* A small number of routes without an explicit type can and do also define properties that do not exist on `Route`. This currently works because they still conform to the `Route` interface and will satisfy TypeScript when passed to functions. However, they will not satisfy TS when adding the explicit `Route` type to the variable declaration. The migration will create code that won't compile for these cases (less than 5 occurrences in g3). For these, the developer can instead use `const x = {pathMatch: 'full' as 'full'}` or `const x = {pathMatch: 'full'} as const`. This could be the approach used by the migration, but I still believe `const x: Route = ...` is more desirable in all cases where this issue does not happen.
* The algorithm will miss assignments like `const route = condition ? [] : {pathMatch: 'full', path: ''}` since it only walks up from the property until it is not longer in the object/routes array. If that node is not a variable declaration, the migration does not attempt a transform. This only happened 1 time in all of google3.
* The algorithm is not entirely optimized. It walks each child until finding the `pathMatch` property, then walks back up to find the variable declaration. It also doesn't short-circuit further processing of other objects in the `Routes` array. I didn't really attempt to optimize this since migrations are one-offs and I didn't feel it caused exceedingly poor performance. I am certainly open to suggestions/improvements here.

### CODE REVIEW

**Comment by atscott** [comment]
[TGP](https://fusion2.corp.google.com/presubmit/tap/427810197/OCL:427810197:BASE:429392582:1645134457303:8012f5d2;filter=;groups=NewlyBroken,PossiblyNewlyFailing/targets) (with a few outstanding fixes)

File: packages/core/schematics/migrations/path-match-type/util.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      } else {
+        node.forEachChild(visitNode);
+      }
+    });
+
+    function visitObjectLiteral(obj: ts.ObjectLiteralExpression, typeChecker: ts.TypeChecker) {
+      const hasPathMatch = obj.properties.some(p => isPropertyWithName(p, 'pathMatch'));
+      const hasPath = obj.properties.some(p => isPropertyWithName(p, 'path'));
+      const childrenProperty = obj.properties.find(p => isPropertyWithName(p, 'children'));
+      // If the object must have _both_ pathMatch _and_ path for us to be reasonably sure that it's
```
Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:34]:
```suggestion
      // The object must have _both_ pathMatch _and_ path for us to be reasonably sure that it's
```
```suggestion
      // If the object must have _both_ pathMatch _and_ path for us to be reasonably sure that it's
```

File: packages/core/schematics/migrations/path-match-type/util.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    sourceFile: ts.SourceFile, typeChecker: ts.TypeChecker, importManager: ImportManager) {
+  const migratedNodesMap: Map<ts.Node, ts.Node> = new Map();
+  (() => {
+    sourceFile.forEachChild(function visitNode(node: ts.Node) {
+      if (isRouteOrRoutesVariableDeclaration(node)) {
+        // The variable declaration is already explicitly typed as `Route` or `Routes` so it does
+        // not need a migration.
+        return;
+      } else if (ts.isObjectLiteralExpression(node)) {
+        visitObjectLiteral(node, typeChecker);
```
Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:24]: Optional idea: Before going down into object literals, should we just track of variable declarations? That way we could skip other non-variable literals and wouldn't need to walk up later (mostly for this to be slightly more idea --not sure about perf necessarily being a big deal as you said). e.g.

```ts
function forEachChild(node) {
  if (isVariableDeclaration(node)) {
    _currentVariableDecl = node;
    visitChildren(node);
    _currentVariableDecl = null;
  } else if (isObjectLiteral(node)) {
    if (_currentVariableDecl !== null && doesNotHaveExplicitType(_currentVariableDecl)) {
      visitObjectLiteral(literal, _currentVariableDecl);
    }
  } else {
     visitChildren(node);
  }
}
```

Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:24]: Optional idea: Before going down into object literals, should we just track of variable declarations? That way we could skip other non-variable literals and wouldn't need to walk up later (mostly for this to be slightly more idea --not sure about perf necessarily being a big deal as you said). e.g.

```ts
function forEachChild(node) {
  if (isVariableDeclaration(node)) {
    _currentVariableDecl = node;
    visitChildren(node);
    _currentVariableDecl = null;
  } else if (isObjectLiteral(node) && _currentVariableDecl !== null) {
   visitObjectLiteral(literal, _currentVariableDecl);
  }
}
```

File: packages/core/schematics/migrations/path-match-type/util.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        if (ts.isArrayLiteralExpression(node)) {
+          typeToUse = importManager.addImportToSourceFile(
+                          sourceFile, 'Routes', '@angular/router') as unknown as ts.TypeNode;
+        } else {
+          typeToUse = importManager.addImportToSourceFile(sourceFile, 'Route', '@angular/router') as
+              unknown as ts.TypeNode;
+        }
+
+        const migrated = ts.factory.updateVariableDeclaration(
+            parent, parent.name, parent.exclamationToken, typeToUse, parent.initializer);
```
Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:82]: Usually we only migrate nodes by limiting the replacement to the smallest possible range, trying to not re-print the whole node as this messes up formatting etc. Is this the case here as well? 

I could imagine us re-formatting the whole object unnecessarily. We should be able to just append the type directly after the variabe name?

Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:82]: Usually we only migrate nodes by limiting the replacement to the smallest possible range, trying to not re-print the whole node as this messes up formatting etc. Is this the case here as well? 

I could imagine us re-formatting the whole object unnecessarily. We should be able to just append the type directly after the variabe name?

File: packages/core/schematics/test/google3/patch_match_type_spec.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    writeFile('tsconfig.json', JSON.stringify({
+      compilerOptions: {
+        module: 'es2015',
+        baseUrl: './',
+        paths: {
+          '@angular/router': ['router.d.ts'],
+          'rxjs': ['rxjs.d.ts'],
+          'rxjs/operators': ['operators.d.ts'],
+        }
```
Comment (devversion) [inline packages/core/schematics/test/google3/patch_match_type_spec.ts:43]: Are these path mappings and the RxJS types needed?

Comment (devversion) [inline packages/core/schematics/test/google3/patch_match_type_spec.ts:43]: Are these path mappings and the RxJS types needed?

File: packages/core/schematics/migrations/path-match-type/util.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      // a `Route` definition.
+      if (hasPath && hasPathMatch) {
+        updateTypeOfParentIfNeeded(obj);
+      } else if (
+          childrenProperty !== undefined && ts.isPropertyAssignment(childrenProperty) &&
+          ts.isArrayLiteralExpression(childrenProperty.initializer)) {
+        // Also need to check the children if it exists
+        for (const child of childrenProperty.initializer.elements) {
+          if (ts.isObjectLiteralExpression(child)) {
+            visitObjectLiteral(child, typeChecker);
```
Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:44]: Nit: is the type checker actually used here?

Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:44]: Nit: is the type checker actually used here?

File: packages/core/schematics/migrations/path-match-type/util.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/schematics/migrations/path-match-type/util.ts:34]: Related to that, just an idea: would it make sense just always adding `Route(s)` if we detect a potential route? not necessarily when it currently has `pathMatch`? that would prepare users for when they start using `pathMatch` and generally seems like a good practice. Feel free to just resolve if you think this is not worth it.

[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/schematics/migrations/path-match-type/util.ts:34]: Related to that, just an idea: would it make sense just always adding `Route(s)` if we detect a potential route? not necessarily when it currently has `pathMatch`? that would prepare users for when they start using `pathMatch` and generally seems like a good practice. Feel free to just resolve if you think this is not worth it.

**Review by devversion — COMMENTED** [review]
Overall looks great, just some comments

File: packages/core/schematics/migrations/path-match-type/util.ts (hunk 2 above)
Comment (atscott) [reply packages/core/schematics/migrations/path-match-type/util.ts:24]: 👍  Thanks for the suggestion.

File: packages/core/schematics/migrations/path-match-type/util.ts (hunk 3 above)
Comment (atscott) [reply packages/core/schematics/migrations/path-match-type/util.ts:82]: I thought we generally updated the most minimal bit using the ts factory API. For this case, unfortunately it's the whole `VariableDeclaration`. I've updated the code to just append the type after the original variable declaration name. Let me know if that looks alright to you.

File: packages/core/schematics/migrations/path-match-type/util.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      // a `Route` definition.
+      if (hasPath && hasPathMatch) {
+        updateCurrentVariableDeclaration();
+      } else if (
+          childrenProperty !== undefined && ts.isPropertyAssignment(childrenProperty) &&
+          ts.isArrayLiteralExpression(childrenProperty.initializer)) {
+        // Also need to check the children if it exists
+        for (const child of childrenProperty.initializer.elements) {
+          if (ts.isObjectLiteralExpression(child)) {
+            visitObjectLiteral(child);
```
Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/util.ts:51]: Can we exit early if we migrate something? probably doesn't matter due to the map?

**Review by devversion — APPROVED** [review]
LGTM

File: packages/core/schematics/migrations/path-match-type/index.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  /** Gets the update recorder for the specified source file. */
+  function getUpdateRecorder(sourceFile: ts.SourceFile): UpdateRecorder {
+    if (updateRecorders.has(sourceFile)) {
+      return updateRecorders.get(sourceFile)!;
+    }
+    const printer = ts.createPrinter();
+    const treeRecorder = tree.beginUpdate(relative(basePath, sourceFile.fileName));
+    const recorder: UpdateRecorder = {
+      updateNode(oldExpr: ts.VariableDeclaration, newExpr: ts.VariableDeclaration) {
+        // treeRecorder.remove(oldExpr.getStart(), oldExpr.getWidth());
```
Comment (devversion) [inline packages/core/schematics/migrations/path-match-type/index.ts:64]: Just saw this: might remove it? seems like a leftover

Comment (atscott) [reply packages/core/schematics/migrations/path-match-type/index.ts:64]: Oh, yep. :)

File: packages/core/schematics/migrations/path-match-type/util.ts (hunk 6 above)
Comment (atscott) [reply packages/core/schematics/migrations/path-match-type/util.ts:51]: 👍  Added a check of the map after the `visit` to exit early.

**Comment by atscott** [comment]
merge assistance: BUILD files for migrations aren't synced to g3 so you'll have to include [this patch](http://cl/439365364) when syncing

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit d56a537196d2ab0f3f48821889020d2f5bc621ba.

**Pull request closed without merging**